WalletWasabi/WabiSabi/WabiSabiCoordinator.cs

```diff
@@ -122,45 +121,50 @@ public void BanDescendant(object? sender, Block block)
 
 	public void BanDoubleSpenders(object? sender, Transaction tx)
 	{
-		bool IsWasabiCoinjoin(uint256 txId)
+		var txId = tx.GetHash();
+		if (IsWasabiCoinJoin(txId, tx))
 		{
-			var finishedCoinJoinIds = Arena.RoundStates
-				.Select(x => x.CoinjoinState)
-				.OfType<SigningState>()
-				.Where(x => x.IsFullySigned)
-				.Select(x => x.CreateUnsignedTransaction().GetHash());
-
-			return CoinJoinIdStore.Contains(txId) || finishedCoinJoinIds.Contains(txId);
+			return;
 		}
 
-		// We don't ban our own coinjoin outputs.
-		if (IsWasabiCoinjoin(tx.GetHash()))
+		var inputOutPoints = tx.Inputs.Select(x => x.PrevOut);
+		var disruptedRounds = Arena.GetRoundsContainingOutpoints(inputOutPoints);
+
+		// No round was disrupted by the received transaction. Nothing to do here.
+		if (disruptedRounds.Length == 0)
 		{
 			return;
 		}
 
-		var outPoints = tx.Inputs.Select(x => x.PrevOut);
-
-		// Detect and punish double spending coins
-		var disrupters = Arena.RoundStates
-			.Where(r => r.Phase != Phase.Ended)
-			.SelectMany(r => r.CoinjoinState.Inputs.Select(a => (RoundId: r.Id, Coin: a)))
-			.Where(x => outPoints.Any(outpoint => outpoint == x.Coin.Outpoint))
-			.ToArray();
-
-		foreach (var (roundId, offender) in disrupters)
+		// Ban all outputs created by the the received transaction because it has spent coins participating in coinjoin rounds.
+		foreach (var indexedOutput in tx.Outputs.AsIndexedOutputs())
 		{
-			Warden.Prison.DoubleSpent(offender.Outpoint, offender.Amount, roundId);
+			Warden.Prison.DoubleSpent(new OutPoint(tx, indexedOutput.N), indexedOutput.TxOut.Value, disruptedRounds);
 		}
 
 		// Abort disrupted rounds
-		var disruptedRounds = disrupters.Select(x => x.RoundId).Distinct();
 		foreach (var roundId in disruptedRounds)
 		{
 			Arena.AbortRound(roundId);
 		}
 	}
 
+	private bool IsWasabiCoinJoin(uint256 txId, Transaction tx) =>
+		CoinJoinIdStore.Contains(txId) || IsFinishedCoinJoin(txId) || IsWasabiCoinJoinLookingTx(tx);
+
+	private bool IsFinishedCoinJoin(uint256 txId) =>
+		Arena.RoundStates
+		.Select(x => x.CoinjoinState)
+		.OfType<SigningState>()
+		.Any(x => x.CreateUnsignedTransaction().GetHash() == txId);
+
+	private bool IsWasabiCoinJoinLookingTx(Transaction tx) =>
+		tx.RBF == false
+		&& tx.Inputs.Count >= Config.MinInputCountByRound
+		&& tx.Inputs.Count <= Config.MaxInputCountByRound
+		&& tx.Outputs.All(x => Config.AllowedOutputTypes.Any(y => x.ScriptPubKey.IsScriptType(y)))
+		&& tx.Outputs.Zip(tx.Outputs.Skip(1), (a, b) => (First: a.Value, Second: b.Value)).All(p => p.First >= p.Second);
```

Comment:
Are these heuristics exactly the same than Dumplings @Szpoti ?

Reply:
No, this is the logic in dumplings:
```cs
    tx.Inputs.All(x => x.PrevOutput.ScriptPubKey.IsScriptType(ScriptType.P2WPKH) || x.PrevOutput.ScriptPubKey.IsScriptType(ScriptType.Taproot))
    && inputCount >= 50
    && inputValues.SequenceEqual(inputValues.OrderByDescending(x => x)) 
    && outputValues.SequenceEqual(outputValues.OrderByDescending(x => x)) 
    && outputValues.Count(x => Wasabi2Denominations.Contains(x.Satoshi)) > outputCount * 0.8;
```

Reply:
Take into account that dumplings, in general, knows less than wasabi. Dumplings assumes the `inputCount` is greater than 50 because it doesn't know what the `MinInputCountByRound` really is. It assumes all outputs are wpkh or tr because it doesn't know what the `AllowedOutputTypes` really are.

OTOH, dumplings know the prevOutputs so, it knows the input values and then it can check the order of the inputs, wasabi cannot. For this same reason it could check that all the inputs are wpkh or tr.

The 80% of std denominations is just a bug waiting to be found. Where is that 80% comes from? If tomorrow we implement pay in coinjoins that would need to change.